File: 205-isomorphic-strings/isomorphic-strings.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {

        if (s.length() != t.length())
            return false;

        unordered_map<char, char> smap;
        unordered_map<char, char> tmap;

        for (int i = 0; i < s.length(); i++) {
            char ch = s[i];
            char th = t[i];

            if (smap.find(ch) != smap.end()) {
                if (smap[ch] != th)
                    return false;
            } else if (tmap.find(th) != tmap.end()) {

                if (tmap[th] != ch)
                    return false;
            }

            smap[ch] = th;
            tmap[th] = ch;
        }
        return true;
    }
};

File: 205-isomorphic-strings/isomorphic-strings.java
class Solution {
    public boolean isIsomorphic(String s, String t) {
        int slen = s.length();
        int tlen = t.length();

        if (slen != tlen)
            return false;

        HashMap<Character, Character> smap = new HashMap<>();
        HashMap<Character, Character> tmap = new HashMap<>();

        for (int i = 0; i < slen; i++) {
            char sval = s.charAt(i);
            char tval = t.charAt(i);

            if (smap.containsKey(sval)) {
                if (smap.get(sval) != tval)
                    return false;
            } else if (tmap.containsKey(tval)) {
                if (tmap.get(tval) != sval)
                    return false;
            }

            smap.put(sval, tval);
            tmap.put(tval, sval);
        }
        return true;

    }
}